src/librbd/io/AioCompletion.h

```diff
@@ -45,7 +45,8 @@ struct AioCompletion {
   } aio_state_t;
 
   mutable std::mutex lock;
-  std::condition_variable cond;
+  std::unique_ptr<std::condition_variable> cond;
```

Comment:
> Modify the conditional variable to be a unique_ptr so that we only allocate memory when it is actually needed (rarely since it only appears to be used in test code).

`AioCompletion::wait_for_complete()` implements the public API (`rbd_aio_wait_for_complete()` and the corresponding C++ variant), so we don't really control its usage.  In-tree, apart from test code, it's used in the implementation of `rbd ls` command (see `src/tools/rbd/action/List.cc`).  In general, I'd agree with the "rarely" estimate.

Instead of involving a dynamic allocation, I wonder if we can get rid of `std::condition_variable` and the associated locking altogether by utilizing `wait()` and `notify_all()` methods directly on `state` (new in C++20):

https://en.cppreference.com/w/cpp/atomic/atomic/wait
https://en.cppreference.com/w/cpp/atomic/atomic/notify_all

Reply:
Sorry, that was poorly worded as I knew about the usage in `rbd ls` as well! I wasn't aware of the public API usage though so thank you for bringing that to my attention.

Replacing the conditional variable with the new C++20 atomic wait/notify functions is a great idea, I will try that and take some more measurements to see if it gives us any further performance improvements.